O365DataRetriever.ps1

```diff
@@ -732,20 +753,45 @@ $ConnectButton.Add_Click( {
 		Write-Host "Connecting to Exchange Online..." -ForegroundColor Cyan
 
 		#If user is MFA enabled
-		$UserAuthN = Get-MsolUser -UserPrincipalName ($creds).UserName 
-		if(($UserAuthN.StrongAuthenticationMethods) -ne $null){
-			Write-Host "You are MFA enabled..." -ForegroundColor Magenta
-			Write-Host "Unfortunately, we don't support users enabled for MFA at this point in time. Try again with a user account not MFA enabled." -ForegroundColor White -BackgroundColor DarkRed
-			$Window.Close()
-			break
+        $UserAuthN = Get-MsolUser -UserPrincipalName ($creds).UserName
+        $userIsMFA = $null -ne $UserAuthN.StrongAuthenticationMethods
```

Comment:
Running the tool with a **non-enabled** MFA user tells me that I am...

Reply:
There seem to be a general Misconception of the settings in StrongAuth* of MSOL Cmdlets. I try to explain and maybe we can confirm (because i'm also not 100% sure about my knowledge)
![image](https://user-images.githubusercontent.com/16267927/44802659-9f52b880-abbc-11e8-8472-270025c7a228.png)

*Property: StrongAuthenticationMethods*
Seem to control if the user has enrolled with MFA. Basically - configured any option to do MFA
Object: 
```
ExtensionData                                    IsDefault MethodType
-------------                                    --------- ----------
System.Runtime.Serialization.ExtensionDataObject      True PhoneAppNotification
System.Runtime.Serialization.ExtensionDataObject     False OneWaySMS
System.Runtime.Serialization.ExtensionDataObject     False PhoneAppOTP
System.Runtime.Serialization.ExtensionDataObject     False TwoWayVoiceMobile
```

*Property: StrongAuthenticationRequirements*
Seem to control the Relying Party trust which are enabled for MFA. (the real - enabled settings)
Object: 
```
ExtensionData                                    RelyingParty RememberDevicesNotIssuedBefore State
-------------                                    ------------ ------------------------------ -----
System.Runtime.Serialization.ExtensionDataObject *            16/04/2018 15:08:14            Enforced
```

*Property: StrongAuthenticationUserDetails*
Seem to save parameters like phone number, pin and email addresses.
```
ExtensionData          : System.Runtime.Serialization.ExtensionDataObject
AlternativePhoneNumber :
Email                  :
OldPin                 :
PhoneNumber            : +43 xxxxxxxxx
Pin                    :
```

so - I guess we have to change the MFA Detection of the user to something like: 

```
(Get-MsolUser -SearchString xxxx | Select-Object -ExpandProperty StrongAuthenticationRequirements).state -eq 'enforced'
```

Thoughts?

Reply:
Before publishing the tool, I tried to make it available for MFA users, simply by checking if the _StrongAuthenticationRequirements_ property on the username wasn't empty. But, MSFT is using different modules / consoles for EXO & Compliance Center _(as an example)_. I don't think there's a need to use "enforced" for the tool.

Reply:
The problem is a bad Microsoft resource. :-)

I think the right answer is the below. I've tested it with both MFA enabled and MFA not-enabled users.

```
$userIsMfa = $false
if( $UserAuthN.StrongAuthenticationMethods -and $UserAuthN.StrongAuthenticationMethods.Count -gt 0 )
{
    $userIsMfa  = $true
}
```
I haven't had a chance to address @VeronicaGeek 's other comment. I'll fix this tonight.

Reply:
What i've found was: [How to require two-step verification for a user](https://docs.microsoft.com/en-us/azure/active-directory/authentication/howto-mfa-userstates)

I think this explains it from a powershell perspective:
```
$st = New-Object -TypeName Microsoft.Online.Administration.StrongAuthenticationRequirement
$st.RelyingParty = "*"
$st.State = "Enabled"
$sta = @($st)
Set-MsolUser -UserPrincipalName bsimon@contoso.com -StrongAuthenticationRequirements $sta
```
this sets the users state to 'enabled', which we can check afterwards. But we have to consider already enabled users with the state 'enforced' which seem to be a second option for the state property.

Reply:
This problem is caused by a change in Get-MsolUser, and probably the graph API itself. Get-MsolUser used to return $null when a list object was empty. Now it returns an empty list. This is why the old method of checking against $null (which almost all sample code does) no longer works and why my "new code" does both checks (verifies that the list isn't $null and that it contains entries).

Will checking StrongAuthenticationRequirements also work? Yes, I believe so, but it's needlessly complex and I don't see any way to tie a particular auth requirement to a particular auth method. This is needed when, for example, a tenant is using both Okta MFA and Azure MFA; for a particular user one or both can be enabled or disabled or enforced.

Reply:
Fixed in latest pull.